src/diag.rs

```diff
@@ -129,84 +133,93 @@ impl From<io::Error> for Diagnostic {
-    sset: &SegmentSet,
+pub(crate) fn to_annotations<'a>(
+    sset: &'a SegmentSet,
+    lc: &mut LineCache,
     mut diags: Vec<(StatementAddress, Diagnostic)>,
-) -> Vec<Notation> {
+) -> Vec<Snippet<'a>> {
     diags.sort_by(|x, y| sset.order.cmp(&x.0, &y.0));
     let mut out = Vec::new();
     for (saddr, diag) in diags {
-        annotate_diagnostic(&mut out, sset, sset.statement(saddr), &diag);
+        out.push(annotate_diagnostic(sset, sset.statement(saddr), lc, &diag));
     }
     out
 }
 
-fn annotate_diagnostic(
-    notes: &mut Vec<Notation>,
-    sset: &SegmentSet,
-    stmt: StatementRef<'_>,
-    diag: &Diagnostic,
-) {
-    struct AnnInfo<'a> {
-        notes: &'a mut Vec<Notation>,
-        sset: &'a SegmentSet,
-        stmt: StatementRef<'a>,
-        level: Level,
-        s: &'static str,
-        args: Vec<(&'static str, String)>,
+/// Converts a raw diagnostics to a snippet
+#[must_use]
+pub fn make_snippet(
+    slices: Vec<Slice<'_>>,
+) -> Snippet<'_> {
+    Snippet {
+        title: Some(Annotation {
+            label: None,
+            id: None,
+            annotation_type: slices[0].annotations[0].annotation_type,
+        }),
+        footer: vec![],
+        slices,
+        opt: FormatOptions {
+            color: true,
+            ..FormatOptions::default()
+        },
     }
+}
 
-    fn ann(info: &mut AnnInfo<'_>, mut span: Span) {
-        if span.is_null() {
-            span = info.stmt.span();
-        }
-        info.notes.push(Notation {
-            source: info.sset.source_info(info.stmt.segment().id).clone(),
-            message: info.s,
-            span,
-            level: info.level,
-            args: mem::take(&mut info.args),
-        })
+/// Creates a `Slice` containing a diagnostic annotation.
+/// 
+/// # Arguments
+///
+/// * `source` - Reference to source data, including the filename and text which could be
+/// used to calculate line numbers or print an invalid excerpt.
+/// * `message` - A message for the diagnostic, which may contain `{placeholders}`.  The
+/// message will be in English but, being not dynamically generated, it is
+/// suitable for remapping with a resource file.
+/// * `span` - The location of the error (byte offset within the SourceInfo; _this is
+/// not the same as the byte offset in the file_).
+/// * `annotation_type` -  Severity level of the message
+/// * `args` - Values to substitute for the `{placeholders}` in the message.  `String`
+/// could be replaced with a richer enum.
+/// * `lc` - A helper struct for keeping track of line numbers in the source file
+fn make_slice<'a>(
+    source: &'a SourceInfo,
+    message: &'static str,
+    span: Span,
+    annotation_type: AnnotationType,
+    args: Vec<(&'static str, String)>,
+    lc: &mut LineCache,
+) -> Slice<'a> {
+    let mut vars = HashMap::new();
+    for (name, value) in args { vars.insert(name.to_string(), value); }
+    let label = strfmt(&message, &vars).unwrap().as_str();
```

Comment:
Basically we need to return a `&'a str` with the error message formatted with the provided arguments, however there is nowhere to construct this string. I assume I will have to provide pre-allocated buffers for that.
Any better idea? @digama0

Reply:
The usual trick I use for returning data that borrows from locals is to convert the function to continuation passing style: that is, instead of returning the data, you pass a function that will receive the data, and call it inside your own call frame. The caller has to pass a closure and cannot use the data once the closure finishes, but they can print the data or copy it into something with a longer lifetime. See for example https://github.com/digama0/mm0/blob/master/mm0-rs/src/compiler.rs#L273-L290 .